File: apple/MiniXiangqi/Settings/Preferences.swift
// The preferences the app keeps, and the one place each of them is read.
//
// docs/product.md, "Product navigation": Settings holds the persistent user
// preferences, it stores no game data, and they live in each platform's own
// preference system rather than in the shared core. Issue #64's Stage 5 design
// fixes the rest of that contract — the key each preference is stored under, its
// type, and what its absence means — because the screen that writes these keys
// and the renderings that read them are built as separate pieces of work. The
// key is the interface between them, so it is written down once, here.
//
// | Key | Type | Absent means |
// |---|---|---|
// | `sound.enabled` | Bool | on |
// | `haptics.enabled` | Bool | on |
// | `deleteConfirmation.enabled` | Bool | on |
// | `placementConfirmation.enabled` | Bool | off |
// | `pieces.symbols` | `hanzi` \| `icons` | `hanzi` |
// | `notation.style` | `traditional` \| `wxf` | the interface language's own |
// | `defaults.firstMover` | `human-first` \| `ai-first` \| `random` | `human-first` |
// | `defaults.aiLevel` | `fast` \| `standard` \| `deep` | `standard` |
//
// **Every one is read at the moment of use** and nothing is cached at launch, so
// a switch takes effect on the next event rather than on the next run. That was
// already the rule the sound gate was written to; all five share it now, and
// share one reader, so that "absent means the accepted default" is stated in one
// place instead of once per consumer.
//
// One default is no longer a constant: `notation.style`'s is the interface
// language's — WXF in English, the traditional rendering in Chinese — by the
// owner's decision of 2026-07-30. That is why a `Choice`'s default is a closure
// below rather than a string. Nothing about the *storage* changed: the resolved
// answer is what an absent key reads as, and the moment somebody chooses in
// Settings a name is written and the language stops being asked.
//
// What a preference *means* is not this file's business. The two choices are
// stored and read as the names in the table above, and the rendering that reads
// one maps it to its own vocabulary: the notation's own type belongs beside the
// notation, not beside the preference that selects it.

import Foundation

enum Preferences {

    // MARK: - The five

    /// Whether the board is heard: the gate in front of `Feedback`'s heard half.
    static let sound = Flag(key: "sound.enabled", whenAbsent: true)

    /// Whether the board is felt: the gate in front of `Feedback`'s felt half.
    /// Offered on macOS, where the hardware question belongs to the trackpad and
    /// the system performer already answers it per machine, per
    /// docs/interaction-design.md § Sound and haptics.
    static let haptics = Flag(key: "haptics.enabled", whenAbsent: true)

    /// Whether deleting a History record asks first. Off means a deletion is
    /// immediate — and permanent, which is what the accepted footer beside the
    /// switch says out loud.
    static let deleteConfirmation = Flag(key: "deleteConfirmation.enabled",
                                         whenAbsent: true)

    /// Whether placing a stone asks first. **Off by default**, which is the
    /// accepted decision: a tap places the stone, and the pending stone is the
    /// flow a player turns on. On means a tap marks the point instead — see
    /// `Game.placementEffect(ofTapAt:…)`, which owns the grammar — and the mark
    /// is committed by tapping it again.
    ///
    /// It is the delete confirmation's shape and not its default, because the
    /// two protect different things: a deletion cannot be undone, and a stone
    /// placed by mistake is one Undo away wherever it is placed.
    static let placementConfirmation = Flag(key: "placementConfirmation.enabled",
                                            whenAbsent: false)

    /// What the discs carry — the accepted symbol sets of
    /// docs/interaction-design.md § Piece symbols: `hanzi`, the default, or
    /// `icons`. Independent of how a move is written: a learner may want icons on
    /// a board whose move list they are learning to read.
    static let pieceSymbols = Choice(key: "pieces.symbols",
                                     whenAbsent: "hanzi",
                                     names: ["hanzi", "icons"])

    /// How a move is written, in the move list and in the numeral strips that
    /// exist to map it to the board: `traditional` or `wxf`, and where nobody
    /// has chosen, whichever one the interface language reads — see
    /// `NotationStyle.resolved(forInterfaceLanguage:)`, which owns that rule and
    /// the reasoning for it.
    static let notationStyle = Choice(key: "notation.style",
                                      whenAbsent: NotationStyle.resolvedForInterfaceLanguage.rawValue,
                                      names: ["traditional", "wxf"])

    /// Which side a new human-versus-AI game opens on — the value the pre-start
    /// draft is initialized from, never the created game's own. The names are
    /// docs/game-data.md's serialized first-mover vocabulary, so the preference,
    /// the frozen configuration and the archive all say the same three words.
    static let defaultFirstMover = Choice(key: "defaults.firstMover",
                                          whenAbsent: "human-first",
                                          names: ["human-first", "ai-first", "random"])

    /// How long a new human-versus-AI game's opponent thinks. 标准 on a new
    /// installation, per the accepted profiles.
    static let defaultAiLevel = Choice(key: "defaults.aiLevel",
                                       whenAbsent: "standard",
                                       names: ["fast", "standard", "deep"])

    /// The two above as the vocabulary the setup page speaks. A stored name
    /// nothing recognises reads as the accepted default, exactly as `Choice`
    /// says it does — a preference file is editable by hand and read by more
    /// than one frontend, and the page still has to open.
    static func defaultFirstMover(in defaults: UserDefaults = Preferences.defaults)
        -> FirstMoverChoice {
        FirstMoverChoice(name: defaultFirstMover.value(in: defaults)) ?? .humanFirst
    }

    static func defaultAiLevel(in defaults: UserDefaults = Preferences.defaults)
        -> AiLevel {
        AiLevel(name: defaultAiLevel.value(in: defaults)) ?? .standard
    }

    // MARK: - How one is read and written

    /// A preference that is on or off: a switch in Settings, and one of the
    /// accepted defaults until somebody moves it.
    struct Flag {
        let key: String
        /// What an absent key means — the state of every first launch.
        let whenAbsent: Bool

        /// Read at the moment of use, never cached.
        ///
        /// Two kinds of stored value are honoured, and everything else is the
        /// accepted default:
        ///
        /// - **A Bool**, which is what the switch writes — and a 0 or 1 number,
        ///   which is the same answer written by something that had no Bool to
        ///   write, another frontend or a hand-edited file.
        /// - **One of the strings that names a Bool** — `1`, `0`, `YES`, `NO`,
        ///   `true`, `false`, in any case. A debug launch argument is how a UI
        ///   test names a preference state without writing anybody's real
        ///   preferences, and the argument domain delivers `-haptics.enabled 0`
        ///   as the *string* `"0"`: a reader that only cast to `Bool` would
        ///   silently answer the default, and the test would be asserting against
        ///   the state it thought it had replaced.
        ///
        /// **Anything else answers `whenAbsent`, and that direction matters.** A
        /// stored value of the wrong type is a preference nobody set — a
        /// hand-edited file, another frontend's mistake, a key colliding with
        /// something that is not this. `bool(forKey:)` would answer *false* for
        /// almost all of it — `on`, `[1]`, a date, an empty string — and *true*
        /// for any non-zero number, so it is not asked: false is the dangerous
        /// answer here, because on `deleteConfirmation.enabled` it deletes a game
        /// without asking, and true is not an answer anybody gave either. So an
        /// unreadable preference reads as the accepted default, exactly as an
        /// unrecognised name does in `Choice` below, and for the same reason: the
        /// app still has to behave, and the default is what it behaves as.
        func value(in defaults: UserDefaults = Preferences.defaults) -> Bool {
            switch defaults.object(forKey: key) {
            case let flag as Bool: flag
            case let text as String: Self.namedFlags[text.lowercased()] ?? whenAbsent
            default: whenAbsent
            }
        }

        /// The strings that name a Bool, lowercased. `as? Bool` already covers a
        /// genuine Bool and a 0-or-1 number, which is why neither is here.
        private static let namedFlags = ["1": true, "0": false,
                                         "yes": true, "no": false,
                                         "true": true, "false": false]

        /// Written as a Bool and never as a string or a number, so that anything
        /// asking `object(forKey:)` is answered with a Bool.
        func set(_ value: Bool, in defaults: UserDefaults = Preferences.defaults) {
            defaults.set(value, forKey: key)
        }
    }

    /// A preference chosen among a fixed set of names, stored as the name itself.
    /// A name rather than an index, because an index is a position in a list that
    /// a later design may reorder and a name is not.
    ///
    /// **The names are plain strings, and no type is declared for them here.**
    /// Each of these is read by a different piece of work — the disc rendering
    /// reads `pieces.symbols`, the move list and the numeral strips read
    /// `notation.style` — and each of those has its own type for what it draws. A
    /// type declared here as well would be a second vocabulary for the same
    /// choice, to be kept in step by hand; the stored string is the interface, as
    /// issue #64's key table says it is.
    struct Choice {
        let key: String
        /// What an absent key means — the state of every first launch — and what
        /// an unrecognised one means too.
        ///
        /// **Asked rather than stored**, because one of these defaults is not a
        /// constant: `notation.style`'s is the interface language's answer, and a
        /// default captured when this table was built would be a different kind
        /// of thing from a default read at the moment of use, which is what every
        /// other value here is. The other three choices pass a literal and read
        /// exactly as they did.
        let whenAbsent: () -> String
        /// Every name this preference accepts.
        let names: [String]

        init(key: String,
             whenAbsent: @autoclosure @escaping () -> String,
             names: [String]) {
            self.key = key
            self.whenAbsent = whenAbsent
            self.names = names
        }

        /// Read at the moment of use, like every other preference here. A stored
        /// string that names nothing is the default rather than a failure: a
        /// preference file can be edited by hand and is read by more than one
        /// frontend, and the board still has to draw.
        func value(in defaults: UserDefaults = Preferences.defaults) -> String {
            guard let stored = defaults.string(forKey: key), names.contains(stored)
            else { return whenAbsent() }
            return stored
        }

        /// Writes one of `names`. Anything else is a mistake in the caller rather
        /// than a state to store, and it is stored anyway, because a preference
        /// this screen writes and cannot read back is worse than one a test can
        /// see is wrong.
        func set(_ value: String, in defaults: UserDefaults = Preferences.defaults) {
            defaults.set(value, forKey: key)
        }
    }

    // MARK: - Where they are kept

    /// The defaults database every read and write above goes through.
    ///
    /// The standard one, except that a debug launch may name a scratch suite with
    /// `-mxq-defaults-suite <name>`: a UI test that *clicks* a switch writes a
    /// real preference, and the player's own preferences are no more a test
    /// fixture than the player's own games are — the same reason
    /// `-mxq-store-name` exists for the store. A named suite is searched ahead of
    /// the app's own domain and behind the argument domain, so a launch argument
    /// still names a state to read and a click still persists one to re-read.
    ///
    /// Resolved once, because a launch argument cannot change while the app runs;
    /// what is never cached is the *value* of a preference, which is a different
    /// thing from the database it is kept in.
    static let defaults: UserDefaults = {
        #if DEBUG
        if let name = DebugLaunch.argument(after: "-mxq-defaults-suite"),
           let suite = UserDefaults(suiteName: name) {
            return suite
        }
        #endif
        return .standard
    }()
}

File: windows/MiniXiangqi.Play/Play/Preferences.cs
// The persistent Settings defaults, read where the pre-start page needs them.
//
// **This pull request reads them and writes none of them.** The Settings screen
// is the next rung of the ladder and it owns writing; what is here is the two
// values docs/interaction-design.md § Starting and configuring a game says the
// pre-start controls are "initialized afresh from the persistent Settings
// defaults whenever the page is entered", read from the keys the Apple frontend
// already stores them under, with the accepted fallbacks for absence.
//
// The keys and their vocabularies are the interface between the screen that
// writes a preference and the surface that reads one, exactly as
// apple/MiniXiangqi/Settings/Preferences.swift says in as many words. This file
// carries the same two rows of that table:
//
// | Key | Type | Absent means |
// |---|---|---|
// | `defaults.firstMover` | `human-first` \| `ai-first` \| `random` | `human-first` |
// | `defaults.aiLevel` | `fast` \| `standard` \| `deep` | `standard` |
// | `deleteConfirmation.enabled` | Bool | on |
//
// The two choice names are docs/game-data.md's serialized vocabulary, so the
// preference, the frozen configuration and the archive all say the same words —
// and so that a preference written by one frontend reads correctly in the other,
// which is the whole reason the key rather than the storage is the contract.
//
// **删除前确认 joined them with the History destination**, which is the surface
// that reads it: it gates a permanent deletion, and docs/game-data.md's Settings
// placement is explicit that the core never reads a preference and that this one
// in particular "gates a permanent deletion but does not perform it". Its absent
// value is on, and a value this reader cannot make sense of reads as on too —
// **false is the dangerous answer here**, because on this key it deletes a game
// without asking, so anything short of an unambiguous no is treated as a yes.
//
// **Where they are kept on Windows is deliberately the smallest answer that
// works.** A JSON file of string values beside the store, read at the moment of
// use and never cached. The Apple frontend has `UserDefaults`; an unpackaged
// Win32 app has no equivalent it can reach without deciding the packaging
// question this repository has parked, so this reads a file and says so. The
// Settings pull request may keep it or replace it; what it may not do is change
// the key or the vocabulary, because those are what the two frontends share.
//
// Every value is read at the moment of use rather than cached at launch, which
// is the rule the Apple side states and the reason a switch takes effect at the
// next event rather than at the next run.

using System.Text.Json;

namespace MiniXiangqi.Play;

/// <summary>Where a stored preference is read from.</summary>
public interface IPreferenceStore
{
    /// <summary>
    /// The value stored under <paramref name="key"/>, or null where nothing is
    /// stored — which is the state of every first launch, and which every
    /// caller answers with the accepted default rather than with a failure.
    /// </summary>
    string? Read(string key);

    /// <summary>
    /// The flag stored under <paramref name="key"/>, or null where nothing
    /// recognisable is stored.
    ///
    /// A flag is a second read rather than a spelling of the first because a
    /// preferences file can be written by a Settings screen and edited by hand,
    /// and a JSON <c>true</c> and the string <c>"true"</c> are both honest ways
    /// to have written one. What is not a flag at all — a number, an object, a
    /// word neither language of this table knows — reads as absent, and the
    /// caller supplies the accepted default.
    /// </summary>
    bool? ReadFlag(string key) => Preferences.Flag(Read(key));
}

/// <summary>
/// The preferences file: <c>%LOCALAPPDATA%\MiniXiangqi\preferences.json</c>, a
/// flat object of string values. Absent, unreadable, or malformed all read as
/// "nothing is stored" — a preference file can be edited by hand and is read by
/// more than one frontend, and the page still has to open.
/// </summary>
public sealed class FilePreferenceStore : IPreferenceStore
{
    private readonly string _path;

    public FilePreferenceStore()
        : this(Path.Combine(
            Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData),
            "MiniXiangqi",
            "preferences.json"))
    {
    }

    public FilePreferenceStore(string path)
    {
        _path = path;
    }

    public string? Read(string key) => Value(key) is { ValueKind: JsonValueKind.String } value
        ? value.GetString()
        : null;

    /// <summary>
    /// A JSON <c>true</c> or <c>false</c> first, and the string spellings of one
    /// after that — which is what the interface's own default does with whatever
    /// <see cref="Read"/> returned.
    /// </summary>
    public bool? ReadFlag(string key) => Value(key) switch
    {
        { ValueKind: JsonValueKind.True } => true,
        { ValueKind: JsonValueKind.False } => false,
        { ValueKind: JsonValueKind.String } text => Preferences.Flag(text.GetString()),
        _ => null,
    };

    private JsonElement? Value(string key)
    {
        try
        {
            using FileStream file = File.OpenRead(_path);
            using JsonDocument document = JsonDocument.Parse(file);
            return document.RootElement.ValueKind == JsonValueKind.Object
                && document.RootElement.TryGetProperty(key, out JsonElement value)
                ? value.Clone()
                : null;
        }
        catch (Exception failure) when (
            failure is IOException or UnauthorizedAccessException or JsonException)
        {
            return null;
        }
    }
}

/// <summary>A store that holds nothing: every key is absent.</summary>
public sealed class NoPreferences : IPreferenceStore
{
    public static readonly NoPreferences Instance = new();

    public string? Read(string key) => null;
}

public static class Preferences
{
    /// <summary>Which side a new human-versus-AI game opens on.</summary>
    public const string DefaultFirstMoverKey = "defaults.firstMover";

    /// <summary>How long a new human-versus-AI game's opponent thinks.</summary>
    public const string DefaultAiLevelKey = "defaults.aiLevel";

    /// <summary>Whether deleting a History record asks first.</summary>
    public const string DeleteConfirmationKey = "deleteConfirmation.enabled";

    /// <summary>
    /// The string spellings of a flag a hand-edited or foreign-written
    /// preferences file may hold. Anything else is not a flag, and the caller's
    /// accepted default answers instead.
    /// </summary>
    public static bool? Flag(string? text) => text?.ToLowerInvariant() switch
    {
        "1" or "yes" or "true" or "on" => true,
        "0" or "no" or "false" or "off" => false,
        _ => null,
    };

    /// <summary>
    /// 删除前确认, which is enabled by default and enabled again wherever the
    /// stored value cannot be read as a flag. Read at the moment of use, like
    /// every other preference here, so a switch takes effect at the next
    /// deletion rather than at the next run.
    /// </summary>
    public static bool ConfirmsDeletion(IPreferenceStore store) =>
        store.ReadFlag(DeleteConfirmationKey) ?? true;

    /// <summary>
    /// 我先手 on a new installation. A stored name nothing recognises reads as
    /// the accepted default too, for the reason the file store gives.
    /// </summary>
    public static FirstMoverChoice DefaultFirstMover(IPreferenceStore store) =>
        PlayVocabulary.FirstMover(store.Read(DefaultFirstMoverKey)) ?? FirstMoverChoice.HumanFirst;

    /// <summary>标准 on a new installation, per the accepted profiles.</summary>
    public static AiLevel DefaultAiLevel(IPreferenceStore store) =>
        PlayVocabulary.Level(store.Read(DefaultAiLevelKey)) ?? AiLevel.Standard;
}
